src/server/http_api.cc

```diff
@@ -0,0 +1,224 @@
+
+#include "base/logging.h"
+#include "core/flatbuffers.h"
+#include "facade/conn_context.h"
+#include "facade/reply_builder.h"
+#include "server/main_service.h"
+#include "util/http/http_common.h"
+
+namespace dfly {
+using namespace util;
+using namespace std;
+namespace h2 = boost::beast::http;
+using facade::CapturingReplyBuilder;
+
+namespace {
+
+bool IsValidReq(flexbuffers::Reference req) {
+  if (!req.IsVector()) {
+    return false;
+  }
+
+  auto vec = req.AsVector();
+  if (vec.size() == 0) {
+    return false;
+  }
+
+  for (size_t i = 0; i < vec.size(); ++i) {
+    if (!vec[i].IsString()) {
+      return false;
+    }
+  }
+  return true;
+}
+
+// Escape a string so that it is legal to print it in JSON text.
+std::string JsonEscape(string_view input) {
+  auto hex_digit = [](int c) -> char { return c < 10 ? c + '0' : c - 10 + 'a'; };
+
+  string out;
+  out.reserve(input.size() + 2);
+  out.push_back('\"');
+
+  auto* p = reinterpret_cast<const unsigned char*>(input.begin());
+  auto* e = reinterpret_cast<const unsigned char*>(input.end());
+
+  while (p < e) {
+    if (*p == '\\' || *p == '\"') {
+      out.push_back('\\');
+      out.push_back(*p++);
+    } else if (*p <= 0x1f) {
+      switch (*p) {
+        case '\b':
+          out.append("\\b");
+          p++;
+          break;
+        case '\f':
+          out.append("\\f");
+          p++;
+          break;
+        case '\n':
+          out.append("\\n");
+          p++;
+          break;
+        case '\r':
+          out.append("\\r");
+          p++;
+          break;
+        case '\t':
+          out.append("\\t");
+          p++;
+          break;
+        default:
+          // this condition captures non readable chars with value < 32,
+          // so size = 1 byte (e.g control chars).
+          out.append("\\u00");
+          out.push_back(hex_digit((*p & 0xf0) >> 4));
+          out.push_back(hex_digit(*p & 0xf));
+          p++;
+      }
+    } else {
+      out.push_back(*p++);
+    }
+  }
+
+  out.push_back('\"');
+  return out;
+}
+
+struct CaptureVisitor {
+  CaptureVisitor() {
+    str = R"({"result":)";
+  }
+
+  void operator()(monostate) {
+  }
+
+  void operator()(long v) {
+    absl::StrAppend(&str, v);
+  }
+
+  void operator()(double v) {
+    absl::StrAppend(&str, v);
+  }
+
+  void operator()(const CapturingReplyBuilder::SimpleString& ss) {
+    absl::StrAppend(&str, "\"", ss, "\"");
+  }
+
+  void operator()(const CapturingReplyBuilder::BulkString& bs) {
+    absl::StrAppend(&str, JsonEscape(bs));
+  }
+
+  void operator()(CapturingReplyBuilder::Null) {
+    absl::StrAppend(&str, "null");
+  }
+
+  void operator()(CapturingReplyBuilder::Error err) {
+    str = absl::StrCat(R"({"error": ")", err.first);
+  }
+
+  void operator()(facade::OpStatus status) {
+    absl::StrAppend(&str, "\"", facade::StatusToMsg(status), "\"");
+  }
+
+  void operator()(const CapturingReplyBuilder::StrArrPayload& sa) {
+    absl::StrAppend(&str, "not_implemented");
```

Comment:
DCHECK?

Reply:
it's a "valid" behavior since it does not crash the server and returns enough output to a user to notify it's not covered.
I prefer not to DCHECK on states that are dependent on a user input.